### PR TITLE
store: update kvrpc.Cleanup proto and change its behaviour

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

Check the `TTL` on the primary lock to decide the real status of a transaction.
After we [update the `TTL` of a transaction](https://github.com/pingcap/tidb/pull/12177), the TTL information on the secondary lock is not accurate.

### What is changed and how it works?

Before this PR, Cleanup always rollbacks a transaction if it's not committed.
After this PR, Cleanup will not rollback a transaction if the primary lock is active.

### Check List 

Tests 

 - Unit test

Side effects

 - Breaking backward compatibility

The semantics of Cleanup API change.

Release note

 - Write release note for bug-fix or new feature.

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @MyonKeminta @coocood @lysu

File: store/tikv/lock_resolver.go [hunk 1]
```diff
@@ -305,17 +305,31 @@ func (lr *LockResolver) ResolveLocks(bo *Backoffer, locks []*Lock) (msBeforeTxnE
...
+				err = errors.Trace(err)
+				return
+			}
+		} else {
+			// If the lock is valid, the txn may be a pessimistic transaction.
+			msBeforeLockExpired := lr.store.GetOracle().UntilExpired(l.TxnID, status.ttl)
+			if msBeforeLockExpired <= 0 {
+				// The txn is a pessimistic transaction, and it's primary lock will expire soon, but
+				// TxnHeartBeat could update the TTL, so we should not clean up the lock.
+				continue
```
Comment (coocood) [inline store/tikv/lock_resolver.go]: I think it's ok to not back off in this case.

File: store/mockstore/mocktikv/mvcc_leveldb.go [hunk 2]
```diff
@@ -926,19 +926,73 @@ func getTxnCommitInfo(iter *Iterator, expectKey []byte, startTS uint64) (mvccVal
...
+// Cleanup API is depreciated, use CheckTxnStatus instead.
+func (mvcc *MVCCLevelDB) Cleanup(key []byte, startTS, currentTS uint64) error {
 	mvcc.mu.Lock()
 	defer func() {
 		mvcc.mu.Unlock()
 		mvcc.deadlockDetector.CleanUp(startTS)
 	}()
 
 	batch := &leveldb.Batch{}
-	err := rollbackKey(mvcc.db, batch, key, startTS)
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply store/mockstore/mocktikv/mvcc_leveldb.go]: It doesn't worth the effort.
The purpose of this PR is to make the cherry-picking process easier for release-3.0 
I plan to remove the `Cleanup` code after https://github.com/pingcap/tidb/pull/11999
The `CheckTxnStatus` API will replace the `Cleanup`, their difference is that the former also consider `minCommitTS` for the large transaction.

[the remark this answers is not in the document]
Comment (tiancaiamao) [reply store/mockstore/mocktikv/mvcc_leveldb.go]: After we remove the code, there is no duplication

File: store/tikv/lock_resolver.go (hunk 1 above)
Comment (tiancaiamao) [reply store/tikv/lock_resolver.go]: Done

File: store/mockstore/mocktikv/mvcc_leveldb.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (coocood) [reply store/mockstore/mocktikv/mvcc_leveldb.go]: The code will not be removed soon, we need to support CleanUp API for upgrading.

File: store/tikv/lock_resolver.go (hunk 1 above)
Comment (tiancaiamao) [reply store/tikv/lock_resolver.go]: I find the old logic is right ...
if msBeforeLockExpired <= 0, we should call continue
or, we can write the code like this:

```
if msBeforeLockExpired > 0 {
    			if msBeforeTxnExpired == 0 || msBeforeLockExpired < msBeforeTxnExpired {
				msBeforeTxnExpired = msBeforeLockExpired
			}
}
```

@coocood

File: store/mockstore/mocktikv/mvcc_leveldb.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply store/mockstore/mocktikv/mvcc_leveldb.go]: On the master branch, we do not call the `CleanUp` function any more soon.
On the 3.0 branch, there is a `CleanUp` but no `CheckTxnStatus`, so no duplication

File: store/tikv/lock_resolver.go [hunk 3]
```diff
@@ -266,61 +266,85 @@ func (lr *LockResolver) BatchResolveLocks(bo *Backoffer, locks []*Lock, loc Regi
...
 		if msBeforeLockExpired <= 0 {
-			tikvLockResolverCountWithExpired.Inc()
-			expiredLocks = append(expiredLocks, l)
-		} else {
-			if msBeforeTxnExpired == 0 || msBeforeLockExpired < msBeforeTxnExpired {
-				msBeforeTxnExpired = msBeforeLockExpired
-			}
-			tikvLockResolverCountWithNotExpired.Inc()
+			expiredSecondaryLocks = append(expiredSecondaryLocks, l)
 		}
```
Comment (coocood) [inline store/tikv/lock_resolver.go]: Forgot to update `msBeforeTxnExpired`?

Comment (tiancaiamao) [reply store/tikv/lock_resolver.go]: move to https://github.com/pingcap/tidb/pull/12212/files#diff-39b30a313dde4d126e56f1828b6dabdcR306
@coocood

File: store/mockstore/mocktikv/mock_tikv_test.go [hunk 4]
```diff
@@ -556,7 +556,7 @@ func (s *testMockTiKVSuite) TestRollbackAndWriteConflict(c *C) {
 	s.mustPutOK(c, "test", "test2", 5, 8)
 
 	// simulate `getTxnStatus` for txn 2.
-	err := s.store.Cleanup([]byte("test"), 2)
+	err := s.store.Cleanup([]byte("test"), 2, math.MaxUint64)
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply store/mockstore/mocktikv/mock_tikv_test.go]: > simulate `getTxnStatus` for txn 2.

It doesn't rollback the txn here @MyonKeminta

File: store/mockstore/mocktikv/mvcc_leveldb.go [hunk 5]
```diff
@@ -926,19 +926,73 @@ func getTxnCommitInfo(iter *Iterator, expectKey []byte, startTS uint64) (mvccVal
...
+		}
+		ok, err := dec.Decode(iter)
+		if err != nil {
+			return err
+		}
+		// If current transaction's lock exists.
+		if ok && dec.lock.startTS == startTS {
+
+			// If the lock has already outdated, clean up it.
+			if uint64(oracle.ExtractPhysical(dec.lock.startTS))+dec.lock.ttl < uint64(oracle.ExtractPhysical(currentTS)) {
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply store/mockstore/mocktikv/mvcc_leveldb.go]: ok

[the remark this answers is not in the document]
Comment (tiancaiamao) [reply store/mockstore/mocktikv/mvcc_leveldb.go]: Done

**Comment by coocood** [comment]
LGTM

File: store/mockstore/mocktikv/mvcc_leveldb.go [hunk 6]
```diff
@@ -926,19 +932,73 @@ func getTxnCommitInfo(iter *Iterator, expectKey []byte, startTS uint64) (mvccVal
 }
 
 // Cleanup implements the MVCCStore interface.
-func (mvcc *MVCCLevelDB) Cleanup(key []byte, startTS uint64) error {
+// Cleanup API is depreciated, use CheckTxnStatus instead.
```
Comment (youjiali1995) [inline store/mockstore/mocktikv/mvcc_leveldb.go]: s/depreciated/deprecated

File: store/tikv/lock_resolver.go [hunk 7]
```diff
@@ -277,61 +279,114 @@ func (lr *LockResolver) ResolveLocks(bo *Backoffer, locks []*Lock) (msBeforeTxnE
...
 
-		cleanRegions, exists := cleanTxns[l.TxnID]
-		if !exists {
-			cleanRegions = make(map[RegionVerID]struct{})
-			cleanTxns[l.TxnID] = cleanRegions
-		}
-
-		err = lr.resolveLock(bo, l, status, cleanRegions)
 		if err != nil {
```
Comment (youjiali1995) [inline store/tikv/lock_resolver.go]: Duplicate of L293-L297.

File: store/mockstore/mocktikv/mock_tikv_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (MyonKeminta) [reply store/mockstore/mocktikv/mock_tikv_test.go]: Why cleanup with current_ts = MaxUint64 doesn't rollback the txn?

**Comment by tiancaiamao** [comment]
PTAL @coocood @youjiali1995 @MyonKeminta

**Comment by coocood** [comment]
LGTM

File: store/mockstore/mocktikv/mock_tikv_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (MyonKeminta) [reply store/mockstore/mocktikv/mock_tikv_test.go]: Oh sorry. I misunderstood it.

**Pull request merged**